### ISSUE TITLE
IDisposable & InstancePerTestCase: Object created for OneTimeSetUp is not disposed

### ISSUE THREAD

**Issue by mvdfugro** [nunit/nunit#3843]
Hi folks,

While trying to better understand the fixture lifecycle, I noticed that for a test fixture with two tests, the constructor is called three times, but the disposer is only called twice. I used the following code:

```C#
    [TestFixture, FixtureLifeCycle(LifeCycle.InstancePerTestCase)]
    public class SingleUseLifecycle : IDisposable
    {
        private static void Log(object obj, string value)
        {
            File.AppendAllText("c:\\temp\\SingleUseLifecycle.log", $"{DateTime.Now}\t{obj?.GetHashCode()}\t{value}\n");
        }
        
        public SingleUseLifecycle() => Log(this, "Constructor");
        public void Dispose() => Log(this, "Disposer");

        [OneTimeSetUp] public static void OneTimeSetUp() => Log(null, "OneTimeSetUp");
        [OneTimeTearDown] public static void OneTimeTearDown() => Log(null, "OneTimeTearDown");
        
        [SetUp] public void SetUp() => Log(this, "SetUp");
        [TearDown] public void TearDown() => Log(this, "TearDown");
        
        [Test] public void Test1() => Log(this, "Test1");
        [Test] public void Test2() => Log(this, "Test2");
    }
```

Which results in the following log:
```
29-4-2021 21:29:18	51781231	Constructor
29-4-2021 21:29:18		OneTimeSetUp
29-4-2021 21:29:18	463695	Constructor
29-4-2021 21:29:18	463695	SetUp
29-4-2021 21:29:18	463695	Test1
29-4-2021 21:29:18	463695	TearDown
29-4-2021 21:29:18	463695	Disposer
29-4-2021 21:29:18	26430654	Constructor
29-4-2021 21:29:18	26430654	SetUp
29-4-2021 21:29:18	26430654	Test2
29-4-2021 21:29:18	26430654	TearDown
29-4-2021 21:29:18	26430654	Disposer
29-4-2021 21:29:18		OneTimeTearDown
```

Objects 463695 and 26430654 are correctly constructed and disposed, but object 51781231 is not.

Object 51781231 gets constructed to run the OneTimeSetUp; a breakpoint in the constructor of the test fixture shows the following stack trace:
```
   at NUnitLifecycle.SingleUseLifecycle..ctor() in C:\Development\NUnitLifecycle\Class1.cs:line 36
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.ConstructorInfo.Invoke(Object[] parameters)
   at NUnit.Framework.Internal.Reflect.Construct(Type type, Object[] arguments)
   at NUnit.Framework.Internal.TypeWrapper.Construct(Object[] args)
   at NUnit.Framework.Internal.Commands.ConstructFixtureCommand.<.ctor>b__0_0(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.BeforeTestCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformOneTimeSetUp()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(WorkItem work)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(WorkItem work)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.RunnerThreadProc()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

Whereas the two other constructor calls are clearly just before executing a test case:
```
   at NUnitLifecycle.SingleUseLifecycle..ctor() in C:\Development\NUnitLifecycle\Class1.cs:line 36
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.ConstructorInfo.Invoke(Object[] parameters)
   at NUnit.Framework.Internal.Reflect.Construct(Type type, Object[] arguments)
   at NUnit.Framework.Internal.TypeWrapper.Construct(Object[] args)
   at NUnit.Framework.Internal.Commands.FixturePerTestCaseCommand.<>c__DisplayClass0_0.<.ctor>b__0(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.BeforeTestCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.<>c__DisplayClass4_0.<PerformWork>b__0()
   at NUnit.Framework.Internal.ContextUtils.<>c__DisplayClass1_0`1.<DoIsolated>b__0(Object _)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at NUnit.Framework.Internal.ContextUtils.DoIsolated(ContextCallback callback, Object state)
   at NUnit.Framework.Internal.ContextUtils.DoIsolated[T](Func`1 func)
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(WorkItem work)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(WorkItem work)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(WorkItem work)
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren()
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork()
   at NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread()
   at NUnit.Framework.Internal.Execution.WorkItem.Execute()
   at NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.RunnerThreadProc()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

Note that this still happens even if no OneTimeSetUp is present.

I haven't been able to get the NUnit tests running locally (yet ;-)), but I would expect that adapting the `DisposableLifeCycleFixtureInstancePerTestCase`  to simply count the construct calls:

```C#
        public static int ConstructCalls { get; set; }
        
        public DisposableLifeCycleFixtureInstancePerTestCase()
        {
            ConstructCalls++;
        }
```

And then comparing the number of construct calls to the number of dispose calls in `NUnit.Framework.Attributes.LifeCycleAttributeTests.InstancePerTestCaseShouldDisposeForEachConstructorCall` should surface the issue.

**Comment by mvdfugro** [comment]
Update: got the tests to work :-) Pushed the failing test to https://github.com/mvdfugro/nunit/tree/issue/3843 for now.

**Comment by rprouse** [comment]
@mvdfugro can you confirm that you are using the latest 3.13.2 release? Also, when you say you got the tests to work, do you mean that you updated the NUnit tests to exhibit the problem? That is what it looks like in your branch.

**Comment by mvdfugro** [comment]
Hi @rprouse!

Sorry, forgot to add that. Yes, this is using NUnit 3.13.2. 

My last message was indeed not very clear -- I meant to say that a) I figured out how to run the tests through Cake and b) that I was able to create a failing test case:

```
> dotnet cake --target=Test
[...]
[few warnings about slow tests]
6) Failed : NUnit.Framework.Attributes.LifeCycleAttributeTests.InstancePerTestCaseShouldDisposeForEachConstructorCall
  Expected: 4
  But was:  3
   at NUnit.Framework.Attributes.LifeCycleAttributeTests.InstancePerTestCaseShouldDisposeForEachConstructorCall()
```

**Comment by mikkelbu** [comment]
Closing this, as this was solved by #3844
